build-logic/src/main/kotlin/ktlint-kotlin-common.gradle.kts

```diff
@@ -5,22 +5,33 @@ plugins {
     kotlin("jvm")
 }
 
+val projectLibs = extensions.getByType<VersionCatalogsExtension>().named("libs")
+val javaCompilationVersion = JavaLanguageVersion.of(projectLibs.findVersion("java-compilation").get().requiredVersion)
+val javaTargetVersion = JavaLanguageVersion.of(projectLibs.findVersion("java-target").get().requiredVersion)
+
 kotlin {
     // All modules, the CLI included, must have an explicit API
     explicitApi()
+    jvmToolchain(jdkVersion = javaCompilationVersion.asInt())
 }
 
+tasks.withType<JavaCompile>().configureEach {
+    options.release.set(javaTargetVersion.asInt())
+}
 tasks.withType<KotlinCompile>().configureEach {
-    compilerOptions {
-        jvmTarget.set(JvmTarget.JVM_1_8)
-    }
+    compilerOptions.jvmTarget.set(JvmTarget.fromTarget(JavaVersion.toVersion(javaTargetVersion).toString()))
 }
 
-// compileJava task and compileKotlin task jvm target compatibility should be set to the same Java version.
-// For some reason, we fallback from toolchain using, see https://github.com/pinterest/ktlint/pull/1787
-tasks.withType<JavaCompile>().configureEach {
-    sourceCompatibility = JavaVersion.VERSION_1_8.toString()
-    targetCompatibility = JavaVersion.VERSION_1_8.toString()
+val requestedJdkVersion = project.findProperty("testJdkVersion")?.toString()?.toInt()
+if (requestedJdkVersion != null) {
+    tasks.register<Test>("testOnJdk") {
+        javaLauncher = javaToolchains.launcherFor { languageVersion = JavaLanguageVersion.of(requestedJdkVersion) }
+
+        // Copy inputs from normal Test task.
+        val testTask = tasks.test.get()
+        classpath = testTask.classpath
+        testClassesDirs = testTask.testClassesDirs
+    }
 }
```

Comment:
This is my new suggestion. My initial approach eagerly registered tasks for all supported java versions. The new approach rely on command line arguments, so if one wants to run test locally they can run:
```cmd
./gradlew testOnJdk -PtestJdkVersion=18
```
which will attempt to run tests against Java 18. 

The only difference is these tasks can't be run via click in IDE + I'm not aware of any possibility to run again multiple java versions within single gradle invocation

Reply:
Usually, I only build against latest java version on my local machine via IDE click. I have trouble to remember all gradle command :-( 
It might be an idea to document the build commands for ktlint in the documentation. That would not only help me, but also other developers on the project.

Reply:
>I only build against latest java version on my local machine via IDE click.

I personally use IDE only to edit the code, I run all commands via terminal 😅 
If you prefer running tasks via IDE, I restored that behavior, and you should now see all additional tasks available under `verification` subtree:
<img width="222" alt="image" src="https://github.com/pinterest/ktlint/assets/36954793/90f2ea44-042e-4531-98d8-cd3c128a24bb">